### ISSUE TITLE
Some static pods fail to start

### ISSUE THREAD

**Issue by Dingshujie** [kubernetes/kubernetes#116880]
### What happened?

if kubelet start with kube-apiserver not ready，during this period，someone remove static pod files，and kube-apiserver recovery，someone update static pod files, move to dir, find kubelet not restart this static pods

### What did you expect to happen?

All static pods should restart.

### How can we reproduce it (as minimally and precisely as possible)?

1. add a static pod yaml to kubelet manifest dir, and wait pod start
2. kubelet api source not ready(apiserver down or network problem)
3. restart kubelet
4. remove this static pod yaml, wait kubelet print `skipping delete because sources aren't ready yet`
5. recover kube-apiserver
6. update this pod yaml(update image vesion) , move updated pod yaml to kubelet manifest dir

kubelet will not restart new pods

### Anything else we need to know?

we find that podWorker managePodLoop will check whether pod allow start;
``` go
// allowPodStart tries to start the pod and returns true if allowed, otherwise
// it requeues the pod and returns false. If the pod will never be able to start
// because data is missing, or the pod was terminated before start, canEverStart
// is false.
func (p *podWorkers) allowPodStart(pod *v1.Pod) (canStart bool, canEverStart bool) {
	if !kubetypes.IsStaticPod(pod) {
		// TODO: Do we want to allow non-static pods with the same full name?
		// Note that it may disable the force deletion of pods.
		return true, true
	}
	p.podLock.Lock()
	defer p.podLock.Unlock()
	status, ok := p.podSyncStatuses[pod.UID]
	if !ok {
		klog.ErrorS(nil, "Pod sync status does not exist, the worker should not be running", "pod", klog.KObj(pod), "podUID", pod.UID)
		return false, false
	}
	if status.IsTerminationRequested() {
		return false, false
	}
	if !p.allowStaticPodStart(status.fullname, pod.UID) {
		p.workQueue.Enqueue(pod.UID, wait.Jitter(p.backOffPeriod, workerBackOffPeriodJitterFactor))
		status.working = false
		return false, true
	}
	return true, true
}

// allowStaticPodStart tries to start the static pod and returns true if
// 1. there are no other started static pods with the same fullname
// 2. the uid matches that of the first valid static pod waiting to start
func (p *podWorkers) allowStaticPodStart(fullname string, uid types.UID) bool {
	startedUID, started := p.startedStaticPodsByFullname[fullname]
	if started {
		return startedUID == uid
	}

	waitingPods := p.waitingToStartStaticPodsByFullname[fullname]
	// TODO: This is O(N) with respect to the number of updates to static pods
	// with overlapping full names, and ideally would be O(1).
	for i, waitingUID := range waitingPods {
		// has pod already terminated or been deleted?
		status, ok := p.podSyncStatuses[waitingUID]
		if !ok || status.IsTerminationRequested() || status.IsTerminated() {
			continue
		}
		// another pod is next in line
		if waitingUID != uid {
			p.waitingToStartStaticPodsByFullname[fullname] = waitingPods[i:]
			return false
		}
		// we are up next, remove ourselves
		waitingPods = waitingPods[i+1:]
		break
	}
	if len(waitingPods) != 0 {
		p.waitingToStartStaticPodsByFullname[fullname] = waitingPods
	} else {
		delete(p.waitingToStartStaticPodsByFullname, fullname)
	}
	p.startedStaticPodsByFullname[fullname] = uid
	return true
}

```
kubelet will skip handle pod remove event when  apiserver config source not ready; so pod worker not marked this pod as a Termination Requested, so not remove uid from startedStaticPodsByFullname;

when apiserver recovers，move an updated pod manifest, uid changes, so allowStaticPodStart return false, lead podworker not to do sync work.

/sig node
@rphillips  @gjkim42 @smarterclayton

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```

</details>
1.25

### Cloud provider

<details>

</details>
none

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>
centos 7

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by gjkim42** [comment]
I couldn't reproduce this issue. This may be fixed in the latest code.
Could you test this in the master branch?

/triage need-more-information

**Comment by liggitt** [comment]
@smarterclayton @bobbypage is there a project or anything tracking reports of issues with static pods?

**Comment by Dingshujie** [comment]
> 

1. mv nginx.yaml file to manifest dir, wait for pod ready
``` yaml
apiVersion: v1
kind: Pod
metadata:
  name: nginx
  namespace: default
spec:
  containers:
  - image: nginx
    imagePullPolicy: Always
    name: nginx
    resources: {}
  dnsPolicy: ClusterFirst
  priority: 0
  restartPolicy: Always
  hostNetwork: true
  terminationGracePeriodSeconds: 30
```
```
[root@test-nodepool-30105-f3gxz ~]# kubectl get pods
NAME                  READY   STATUS    RESTARTS   AGE
nginx-192.168.0.233   1/1     Running   0          2m27s
```

2. add a reject iptables , 192.168.0.146 is master address
```
iptables -A OUTPUT -d 192.168.0.146 -j REJECT
```
3. restart kubelet
4. remove nginx.yaml from dir

5. delete reject rules 
```
iptables -D OUTPUT 2
```
now use crictl ps, container exist
```
[root@test-nodepool-30105-f3gxz ~]# crictl ps
CONTAINER           IMAGE               CREATED              STATE               NAME                 ATTEMPT             POD ID
0870acc42f8cc       992e3b7be0465       4 minutes ago        Running             nginx                2                   f58e4cc4486a0
```
6. change image to tomcat, move updated nginx.yaml file to manifest dir,
``` yaml
apiVersion: v1
kind: Pod
metadata:
  name: nginx
  namespace: default
spec:
  containers:
  - image: tomcat
    imagePullPolicy: Always
    name: nginx
    resources: {}
  dnsPolicy: ClusterFirst
  priority: 0
  restartPolicy: Always
  hostNetwork: true
  terminationGracePeriodSeconds: 30
```
watch containers status, pod never restart, and change images to tomcat 

use master commit f0791b50143856177878e21bb44beb5e3e36cc78 to reproduce this issue

**Comment by Dingshujie** [comment]
another information:
Moving nginx.yaml out and in of dir can't recover until change yaml to original file.
because static pod uid  calculate based on pod hash.

**Comment by bobbypage** [comment]
I attempted to repro, but didn't see the same behavior. I saw that the nginx pod was actually removed after the api-server came back up.


```
# create a single node cluster from kind. k8s repo is checked out at c3e7eca7fd38454200819b60e58144d5727f1bbc
$ gh gist create  ~/dev_scripts/kind_single_node_master.sh
```

```
$ docker exec -it kind-worker /bin/bash
$ cd /etc/kubernetes/manifests
$ vim nginx.yaml (and copy the manifest from https://github.com/kubernetes/kubernetes/issues/116880#issuecomment-1483686844)

# outside the kind-worker, verify nginx static pod is up
$ kubectl get pods
NAME                READY   STATUS    RESTARTS   AGE
nginx-kind-worker   1/1     Running   0          7s
```

```
$ docker exec -it kind-control-plane /bin/bash
root@kind-control-plane:/# cd /etc/kubernetes/manifests/
root@kind-control-plane:/etc/kubernetes/manifests# mv kube-apiserver.yaml ..

# api server is down
root@kind-control-plane:/etc/kubernetes/manifests# crictl ps
CONTAINER           IMAGE               CREATED             STATE               NAME                      ATTEMPT             POD ID              POD
2dd6a0ef5f8fc       7191aa84d698f       3 seconds ago       Running             local-path-provisioner    4                   de3a7c2a3e6f0       local-path-provisioner-7dbf974f64-6xvrj
1b651519f0dc9       d6e3e26021b60       18 seconds ago      Running             kindnet-cni               1                   8b70a75a6ad60       kindnet-zcr7x
90513474b3d6d       30c5fd2ae2384       19 seconds ago      Running             kube-controller-manager   1                   6c0660cbc56e1       kube-controller-manager-kind-control-plane
b147331152798       99e62fb8e045a       20 seconds ago      Running             kube-scheduler            1                   e34a2f92ef6b9       kube-scheduler-kind-control-plane
d74f97dcccac8       ead0a4a53df89       3 minutes ago       Running             coredns                   0                   50407c2cadbfc       coredns-5d78c9869d-74c9f
9955d6a41c4c6       ead0a4a53df89       3 minutes ago       Running             coredns                   0                   4861c6b6e2d44       coredns-5d78c9869d-btgz5
456037b3139eb       86b6af7dd652c       4 minutes ago       Running             etcd                      0                   2738fbfc03aa9       etcd-kind-control-plane


# on the node
root@kind-worker:/etc/kubernetes/manifests# systemctl restart kubelet.service
# move nginx.yaml out of the static manifests
root@kind-worker:/etc/kubernetes/manifests# mv nginx.yaml ..
```


```
# start backup the api-server
root@kind-control-plane:/etc/kubernetes/manifests# mv ../kube-apiserver.yaml .
```


```
root@kind-worker:/etc/kubernetes/manifests# ls
# ... nginx manifest is removed ...  
root@kind-worker:/etc/kubernetes/manifests# crictl ps
CONTAINER           IMAGE               CREATED             STATE               NAME                ATTEMPT             POD ID              POD
c1517522b9211       d6e3e26021b60       2 minutes ago       Running             kindnet-cni         1                   ceca7d10404fe       kindnet-tmk9q
7e7cb56689fc4       ac232364af842       3 minutes ago       Running             nginx               0                   5ed556814bdee       nginx-kind-worker

... after a few seconds nginx is not running anymore ...

CONTAINER           IMAGE               CREATED             STATE               NAME                ATTEMPT             POD ID              POD
c1517522b9211       d6e3e26021b60       2 minutes ago       Running             kindnet-cni         1                   ceca7d10404fe       kindnet-tmk9q
```

**Comment by bobbypage** [comment]
I tried the same steps on v1.26.2 and was able to repro (nginx was not deleted). This is likely fixed due to https://github.com/kubernetes/kubernetes/pull/113145 which was merged as part of v1.27 cycle because it properly handles terminating orphaned pods.

**Comment by Dingshujie** [comment]
> I tried the same steps on v1.26.2 and was able to repro (nginx was stuck and not terminated).
> 
> This is likely fixed due to #113145 which was merged as part of v1.27 cycle because it properly handles terminating orphaned pods.

take glanced at #113145, indeed solve the problem

**Comment by bobbypage** [comment]
great, thanks. Let's close this issue then since it is fixed in v1.27.

/close
